java/vector/src/test/java/org/apache/arrow/vector/TestListVector.java

```diff
@@ -827,4 +827,51 @@ public void testSetInitialCapacity() {
       assertEquals(1, vector.getDataVector().getValueCapacity());
     }
   }
+
+  @Test
+  public void testClearAndReuse() {
+    try (final ListVector vector = ListVector.empty("list", allocator)) {
+      BigIntVector bigIntVector = (BigIntVector) vector.addOrGetVector(FieldType.nullable(MinorType.BIGINT.getType())).getVector();
+      vector.setInitialCapacity(10);
+      vector.allocateNew();
+
+      vector.startNewValue(0);
+      bigIntVector.setSafe(0, 7);
+      vector.endValue(0, 1);
+      vector.startNewValue(1);
+      bigIntVector.setSafe(1, 8);
+      vector.endValue(1, 1);
+      vector.setValueCount(2);
+
+      Object result = vector.getObject(0);
+      ArrayList<Long> resultSet = (ArrayList<Long>) result;
+      assertEquals(new Long(7), resultSet.get(0));
+
+      result = vector.getObject(1);
+      resultSet = (ArrayList<Long>) result;
+      assertEquals(new Long(8), resultSet.get(0));
+
+      vector.clear();
+
+      // The list vector should reuse a buffer when reallocating the offset buffer
```

Comment:
Curious about this comment - Does `clear` not release the buffer? How is the buffer reused?

Reply:
Yeah, `clear` released the buffer, then next time it goes to realloc and requests a new buffer if the buffer manager still has one available, it will reuse it.  So the case I was seeing was that an old data buffer was being reused for the new `ListVector` offset buffer - causing all kinds of fun output

I'm not sure if there is a better way to force this issue more directly, but this test seemed to cause it.

Reply:
I see.

This is probably a separate discussion but another option is have the buffer manager zero out the buffer before returning to the caller  instead of having the caller do the zero out. It seems the code would be cleaner that way.

@jacques-n @siddharthteotia @BryanCutler what do you think?

Reply:
That's true too. Anyway, this PR looks good to me. Thanks!